Shopify/cli
# Add DeviceAuth flow for Spin

### WHY are these changes introduced?
The use of cloud environments for development has increased recently (both in and out of shopify). 

Running the Shopify CLI inside one of those environments is a bit tricky as the CLI usually depends on localhost being a real localhost, opening a tunnel or creating a local server for the auth process or to serve apps.

This PR fixes the first blocker users are going to encounter in cloud environments: **Authentication**.
Instead of relying on a local server that listens to identity callbacks when the user completes the authentication, we are adding a new (complementary) auth flow that allows you to login in local environment (your local browser) but getting a token in your cloud environment.

This new auth method will only be used if the CLI detects that is being executed in a cloud environment.
By default the classic method will be used.

Fixes https://github.com/Shopify/shopify-cli-planning/issues/330



### WHAT is this pull request doing?
- Add new Device Auth flow to obtain an identity token, once obtained the process of exchanging that token for application tokens or refreshing an expired token is the same.
- Improve how spin FQDN is obtained, specially when the CLI is being executed inside spin.
- Add new env var to force using the new device auth flow
- Detect common cloud environments using pre-defined env vars



### How to test your changes?
- Try the new auth flow using `SHOPIFY_CLI_DEVICE_AUTH=1` (remember to logout first)
- Try this branch in a Spin instance, it should automatically activate the device auth flow.



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist



https://user-images.githubusercontent.com/3757193/189137666-d0ff97a1-be4e-45da-8f3b-206545c3449b.mov
